daffodil-io/src/main/scala/org/apache/daffodil/io/InputSource.scala

```diff
@@ -255,18 +292,27 @@ class BucketingInputSource(
    */
   private def fillBucketsToIndex(goalBucketIndex: Long, bytesNeededInBucket: Long): Boolean = {
     var lastBucketIndex = buckets.length - 1
-    var hasMoreData = true
+
     var needsMoreData = goalBucketIndex > lastBucketIndex || (goalBucketIndex == lastBucketIndex && bytesNeededInBucket > bytesFilledInLastBucket)
 
     while (needsMoreData && hasMoreData) {
       // Try to fill the rest of this bucket, regardless of how many bytes are
       // actually needed
       val emptyBytesInLastBucket = bucketSize - bytesFilledInLastBucket
 
+      Assert.invariant(emptyBytesInLastBucket >= 0)
       // Try to read enough bytes to fill the rest of this bucket. Note that
       // the .read() function could hit EOF (returns -1) or could return
-      // anywhere from zero to emptyBytesInLastBucket bytes
-      val bytesRead = inputStream.read(buckets(lastBucketIndex).bytes, bytesFilledInLastBucket, emptyBytesInLastBucket)
+      // anywhere from 0 to emptyBytesInLastBucket bytes
+      val bytesRead =
+        inputStream.read(
+          buckets(lastBucketIndex).bytes,
+          bytesFilledInLastBucket,
+          emptyBytesInLastBucket)
+
+      // check for bad inputStream behavior. It's not our fault!
+      if (bytesRead == 0 && emptyBytesInLastBucket > 0)
```

Comment:
If we can change the above assert to > 0, this can just become ``bytesRead == 0``